pkg/comp-functions/functions/common/backup/backup.go

```diff
@@ -0,0 +1,267 @@
+	k8upv1 "github.com/k8up-io/k8up/v2/api/v1"
+	"github.com/sethvargo/go-password/password"
+	appcatv1 "github.com/vshn/appcat/v4/apis/v1"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/functions/common"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/ptr"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	credentialSecretName = "backup-bucket-credentials"
+	k8upRepoSecretName   = "k8up-repository-password"
+	k8upRepoSecretKey    = "password"
+	backupScriptCMName   = "backup-script"
+)
+
+// AddK8upBackup creates an S3 bucket and a K8up schedule according to the composition spec.
+func AddK8upBackup(ctx context.Context, svc *runtime.ServiceRuntime, comp common.InfoGetter) error {
+
+	l := controllerruntime.LoggerFrom(ctx)
+
+	l.Info("Creating backup bucket")
+	err := createObjectBucket(ctx, comp, svc)
+	if err != nil {
+		return fmt.Errorf("cannot create backup bucket: %w", err)
+	}
+
+	l.Info("Creating repository password")
+	err = createRepositoryPassword(ctx, comp, svc)
+	if err != nil {
+		return fmt.Errorf("cannot create repository password: %w", err)
+	}
+
+	l.Info("Creating backup schedule")
+	err = createK8upSchedule(ctx, comp, svc)
+	if err != nil {
+		return fmt.Errorf("cannot create backup schedule, %w", err)
+	}
+
+	return nil
+}
+
+func createObjectBucket(ctx context.Context, comp common.InfoGetter, svc *runtime.ServiceRuntime) error {
+
+	ob := &appcatv1.XObjectBucket{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: comp.GetName() + "-backup",
+		},
+		Spec: appcatv1.XObjectBucketSpec{
+			Parameters: appcatv1.ObjectBucketParameters{
+				BucketName: comp.GetName() + "-backup",
+				Region:     svc.Config.Data["bucketRegion"],
+			},
+			ResourceSpec: xpv1.ResourceSpec{
+				WriteConnectionSecretToReference: &xpv1.SecretReference{
+					Namespace: comp.GetInstanceNamespace(),
+					Name:      credentialSecretName,
+				},
+			},
+		},
+	}
+
+	return svc.SetDesiredComposedResource(ob)
+}
+
+func createRepositoryPassword(ctx context.Context, comp common.InfoGetter, svc *runtime.ServiceRuntime) error {
+
+	l := controllerruntime.LoggerFrom(ctx)
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      k8upRepoSecretName,
+			Namespace: comp.GetInstanceNamespace(),
+		},
+	}
+
+	err := svc.GetObservedKubeObject(secret, comp.GetName()+"-k8up-repo-pw")
+	if err != nil && err != runtime.ErrNotFound {
+		return err
+	}
+
+	if _, ok := secret.Data[k8upRepoSecretKey]; ok {
+		l.V(1).Info("secret is not empty")
+		return svc.SetDesiredKubeObject(secret, comp.GetName()+"-k8up-repo-pw")
+	}
+
+	pw, err := password.Generate(64, 5, 5, false, true)
+	if err != nil {
+		return err
+	}
+
+	secret.Data = map[string][]byte{
+		k8upRepoSecretKey: []byte(pw),
+	}
+
+	return svc.SetDesiredKubeObject(secret, comp.GetName()+"-k8up-repo-pw")
```

Comment:
Why are you not using the `AddCredentialsSecret` function you created for this purpose?

Reply:
I just refactored the code without changing too much of the logic.

I like the idea, this way we can save one more object.